packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -399,6 +387,33 @@ export class ApolloServerBase {
     this.graphqlPath = path;
   }
 
+  /**
+   * Update schema and all data derived from schema, including clearing any caches that might depend on the schema
+   */
+  public updateSchema(schema: GraphQLSchema) {
+    this.schema = schema;
+    this.schemaHash = generateSchemaHash(this.schema);
+
+    if (this.engineServiceId) {
+      const { EngineReportingAgent } = require('apollo-engine-reporting');
+      this.engineReportingAgent = new EngineReportingAgent(
+        typeof this.engine === 'object' ? this.engine : Object.create(null),
+        {
+          schema: this.schema,
+          schemaHash: this.schemaHash,
+          engine: {
+            serviceID: this.engineServiceId,
+          },
+        },
+      );
+      // Let's keep this extension second so it wraps everything, except error formatting
+      this.extensions.push(() => this.engineReportingAgent!.newExtension());
+    }
+
+    this.initializeDocumentStore();
+    this.willStart();
```

Comment:
As far as I can tell the plugins only receive the server's schema via their `.serverWillStart` hook. So it becomes a question of who do we break. 

Maybe add a plugin hook for `.schemaDidChange`? And possibly fallback to `serverWillStart` if `schemaDidChange` is not defined? Don't know what would break fewer setups.